persistence/sql/persister_login.go

```diff
@@ -84,3 +88,123 @@ func (p *Persister) DeleteExpiredLoginFlows(ctx context.Context, expiresAt time.
 	}
 	return nil
 }
+
+func (p *Persister) CreateLoginCode(ctx context.Context, codeParams *code.CreateLoginCodeParams) (*code.LoginCode, error) {
+	ctx, span := p.r.Tracer(ctx).Tracer().Start(ctx, "persistence.sql.CreateLoginCode")
+	defer span.End()
+
+	now := time.Now().UTC()
+	loginCode := &code.LoginCode{
+		IdentityID:  codeParams.IdentityID,
+		Address:     codeParams.Address,
+		AddressType: codeParams.AddressType,
+		CodeHMAC:    p.hmacValue(ctx, codeParams.RawCode),
+		IssuedAt:    now,
+		ExpiresAt:   now.UTC().Add(p.r.Config().SelfServiceCodeMethodLifespan(ctx)),
+		FlowID:      codeParams.FlowID,
+		NID:         p.NetworkID(ctx),
+		ID:          uuid.Nil,
+	}
+
+	if err := p.GetConnection(ctx).Create(loginCode); err != nil {
+		return nil, sqlcon.HandleError(err)
+	}
+	return loginCode, nil
+}
+
+func (p *Persister) UseLoginCode(ctx context.Context, flowID uuid.UUID, identityID uuid.UUID, codeVal string) (*code.LoginCode, error) {
+	ctx, span := p.r.Tracer(ctx).Tracer().Start(ctx, "persistence.sql.UseLoginCode")
+	defer span.End()
+
+	var loginCode *code.LoginCode
+
+	nid := p.NetworkID(ctx)
+	flowTableName := new(login.Flow).TableName(ctx)
+
+	if err := sqlcon.HandleError(p.Transaction(ctx, func(ctx context.Context, tx *pop.Connection) (err error) {
+		//#nosec G201 -- TableName is static
+		if err := sqlcon.HandleError(tx.RawQuery(fmt.Sprintf("UPDATE %s SET submit_count = submit_count + 1 WHERE id = ? AND nid = ?", flowTableName), flowID, nid).Exec()); err != nil {
+			return err
+		}
+
+		var submitCount int
+		// Because MySQL does not support "RETURNING" clauses, but we need the updated `submit_count` later on.
+		//#nosec G201 -- TableName is static
+		if err := sqlcon.HandleError(tx.RawQuery(fmt.Sprintf("SELECT submit_count FROM %s WHERE id = ? AND nid = ?", flowTableName), flowID, nid).First(&submitCount)); err != nil {
+			if errors.Is(err, sqlcon.ErrNoRows) {
+				// Return no error, as that would roll back the transaction
+				return nil
+			}
+			return err
+		}
+
+		if submitCount > 5 {
+			return errors.WithStack(code.ErrCodeSubmittedTooOften)
+		}
+
+		var loginCodes []code.LoginCode
+		if err = sqlcon.HandleError(tx.Where("nid = ? AND selfservice_login_flow_id = ? AND identity_id = ?", nid, flowID, identityID).All(&loginCodes)); err != nil {
```

Comment:
@Benehiko can you please explain why and if the `identity_id` is important in this query please? Are there security implications?

Reply:
Since there are two steps in the login process:
1. submit identifier
2. submit code

In the second submit, we still accept the identifier (again) from the user. Although we do a validation step to ensure that the identifiers match between the first submit and the second, I thought it might be best to also have the login flow attached to the identity from the first submit. 

This ensures that the second submit step has an additional check before getting the login _code. I guess since you would need to have the correct code and the flow id and somehow bypass the check before the persistence layer, it is unlikely that you would be able to circumvent it.